File: estonian-courier-sim/app/src/main/java/ee/kuller/app/ui/screens/OrderScreen.kt
package ee.kuller.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ee.kuller.app.ChatMsg
import ee.kuller.app.GameViewModel
import ee.kuller.app.threadLabel
import ee.kuller.app.ui.SpeakButton

@Composable
fun OrderScreen(vm: GameViewModel, onExit: () -> Unit) {
    val session = vm.session
    if (session == null) { onExit(); return }
    if (session.finished && vm.lastResult != null) { ResultView(vm, onExit); return }

    val messages = session.messages(session.activeThread)
    val showTyping = session.typing == session.activeThread
    val listState = rememberLazyListState()
    LaunchedEffect(messages.size, showTyping, session.awaiting, session.atEnd, session.activeThread, session.nav != null) {
        val extra = if (showTyping) 1 else 0
        if (messages.isNotEmpty() || showTyping) {
            listState.animateScrollToItem(messages.size + 1 + extra)
            listState.animateScrollToItem(messages.size + 1 + extra)
        }
    }

    Column(Modifier.fillMaxSize().background(MaterialTheme.colorScheme.background)) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 8.dp, vertical = 4.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            TextButton(onClick = { vm.cancelOrder(); onExit() }) {
                Icon(Icons.Filled.Close, null, Modifier.size(18.dp)); Text(" Отмена")
            }
            Column(Modifier.weight(1f)) {
                Text("${session.order.restaurant} → ${session.order.customer}",
                    fontWeight = FontWeight.Bold, fontSize = 14.sp,
                    color = MaterialTheme.colorScheme.onBackground)
                Text(session.order.address, fontSize = 11.sp,
                    color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.6f))
            }
        }
        LinearProgressIndicator(progress = { session.progress },
            modifier = Modifier.fillMaxWidth().height(5.dp))

        // Вкладки-чаты
        if (session.tabs.size > 1) {
            Row(
                Modifier.fillMaxWidth().horizontalScroll(rememberScrollState())
                    .padding(horizontal = 8.dp, vertical = 6.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                session.tabs.forEach { t ->
                    FilterChip(
                        selected = t == session.activeThread,
                        onClick = { vm.setActiveThread(t) },
                        label = { Text(threadLabel(t) + if (t in session.unread) "  🔴" else "") }
                    )
                }
            }
        }

        // Лента активного чата
        LazyColumn(
            state = listState,
            modifier = Modifier.weight(1f).fillMaxWidth().padding(horizontal = 12.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            item { Spacer(Modifier.height(8.dp)) }
            itemsIndexed(messages) { _, msg ->
                val ru = if (vm.hideRu) "" else vm.maskRu(msg.ru)
                ChatBubble(msg, ru) { vm.speak(msg.et, msg.fromCourier, msg.thread) }
            }
            if (showTyping) item { TypingBubble(threadLabel(session.activeThread)) }
            item { Spacer(Modifier.height(4.dp)) }
        }

        // Нижняя панель
        Surface(color = MaterialTheme.colorScheme.surface, shadowElevation = 8.dp) {
            Column(Modifier.fillMaxWidth().padding(12.dp)) {
                val ask = session.currentAsk
                val nav = session.nav
                when {
                    session.atEnd -> {
                        Button(onClick = { vm.finishDelivery() },
                            modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(14.dp)) {
                            Text("Завершить доставку 🏁", fontWeight = FontWeight.Bold)
                        }
                    }
                    nav != null -> {
                        Text(nav.promptRu, fontWeight = FontWeight.Bold, fontSize = 15.sp,
                            color = MaterialTheme.colorScheme.onSurface,
                            modifier = Modifier.padding(bottom = 8.dp))
                        nav.options.forEach { opt ->
                            Button(onClick = { vm.navChoose(opt) },
                                modifier = Modifier.fillMaxWidth().padding(vertical = 4.dp),
                                shape = RoundedCornerShape(14.dp)) {
                                Text(opt.labelRu, fontWeight = FontWeight.Bold)
                            }
                        }
                    }
                    session.awaiting && ask != null && ask.thread == session.activeThread -> {
                        // Прокручиваемые варианты + ЗАКРЕПЛЁННАЯ кнопка «Отправить»
                        Column(Modifier.heightIn(max = 270.dp).verticalScroll(rememberScrollState())) {
                            Text(ask.promptRu, fontWeight = FontWeight.Bold, fontSize = 14.sp,
                                color = MaterialTheme.colorScheme.onSurface)
                            Text(
                                if (ask.courier) "🛵 Выберите, что сказать:" else "🛵 Выберите ответ:",
                                fontSize = 12.sp, fontWeight = FontWeight.Bold,
                                color = MaterialTheme.colorScheme.secondary,
                                modifier = Modifier.padding(top = 2.dp, bottom = 6.dp)
                            )
                            ask.choices.forEachIndexed { i, choice ->
                                AnswerOption(
                                    et = choice.et, ru = choice.ru,
                                    selected = session.selected == i,
                                    onClick = { vm.select(i) },
                                    onSpeak = { vm.speak(choice.et, true, ask.thread) }
                                )
                            }
                        }
                        Spacer(Modifier.height(8.dp))
                        Button(
                            onClick = { vm.confirm() },
                            enabled = session.selected != null,
                            modifier = Modifier.fillMaxWidth(),
                            shape = RoundedCornerShape(14.dp)
                        ) {
                            Icon(Icons.AutoMirrored.Filled.Send, null, Modifier.size(18.dp))
                            Text("  Отправить", fontWeight = FontWeight.Bold)
                        }
                    }
                    session.awaiting && ask != null -> {
                        Text("Ответьте в чате «${threadLabel(ask.thread)}»",
                            fontSize = 13.sp, color = MaterialTheme.colorScheme.onSurface,
                            modifier = Modifier.padding(bottom = 8.dp))
                        OutlinedButton(onClick = { vm.setActiveThread(ask.thread) },
                            modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(14.dp)) {
                            Text("Открыть «${threadLabel(ask.thread)}»")
                        }
                    }
                    else -> {
                        Text("💬 …", fontSize = 16.sp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
                            modifier = Modifier.padding(vertical = 12.dp))
                    }
                }
            }
        }
    }
}

@Composable
private fun ChatBubble(msg: ChatMsg, displayRu: String, onSpeak: () -> Unit) {
    val scheme = MaterialTheme.colorScheme
    val courier = msg.fromCourier
    val bubbleColor = if (courier) scheme.primary else scheme.surfaceVariant
    val onBubble = if (courier) scheme.onPrimary else scheme.onSurface
    val shape = RoundedCornerShape(
        topStart = 18.dp, topEnd = 18.dp,
        bottomStart = if (courier) 18.dp else 4.dp,
        bottomEnd = if (courier) 4.dp else 18.dp
    )
    Row(Modifier.fillMaxWidth(),
        horizontalArrangement = if (courier) Arrangement.End else Arrangement.Start) {
        Column(
            horizontalAlignment = if (courier) Alignment.End else Alignment.Start,
            modifier = Modifier.widthIn(max = 290.dp)
        ) {
            Text(msg.label, fontSize = 11.sp, fontWeight = FontWeight.Bold,
                color = scheme.onBackground.copy(alpha = 0.6f),
                modifier = Modifier.padding(start = 6.dp, end = 6.dp, bottom = 2.dp))
            Surface(color = bubbleColor, shape = shape) {
                Row(Modifier.padding(start = 12.dp, end = 4.dp, top = 8.dp, bottom = 8.dp),
                    verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.widthIn(max = 230.dp)) {
                        Text(msg.et, fontWeight = FontWeight.Bold, fontSize = 16.sp, color = onBubble)
                        if (displayRu.isNotBlank()) {
                            Text(displayRu, fontSize = 13.sp, color = onBubble.copy(alpha = 0.8f))
                        }
                    }
                    SpeakButton(onClick = onSpeak, tint = if (courier) scheme.onPrimary else scheme.primary)
                }
            }
        }
    }
}

@Composable
private fun TypingBubble(label: String) {
    val scheme = MaterialTheme.colorScheme
    Column(horizontalAlignment = Alignment.Start) {
        Text(label, fontSize = 11.sp, fontWeight = FontWeight.Bold,
            color = scheme.onBackground.copy(alpha = 0.6f),
            modifier = Modifier.padding(start = 6.dp, end = 6.dp, bottom = 2.dp))
        Surface(color = scheme.surfaceVariant,
            shape = RoundedCornerShape(topStart = 18.dp, topEnd = 18.dp, bottomStart = 4.dp, bottomEnd = 18.dp)) {
            Text("печатает •••", fontSize = 14.sp,
                color = scheme.onSurface.copy(alpha = 0.6f),
                modifier = Modifier.padding(horizontal = 14.dp, vertical = 10.dp))
        }
    }
}

@Composable
private fun AnswerOption(
    et: String, ru: String, selected: Boolean,
    onClick: () -> Unit, onSpeak: () -> Unit,
) {
    val scheme = MaterialTheme.colorScheme
    val border: Color = if (selected) scheme.primary else scheme.onSurface.copy(alpha = 0.12f)
    val bg: Color = if (selected) scheme.primary.copy(alpha = 0.08f) else scheme.surface
    Card(
        modifier = Modifier.fillMaxWidth().padding(vertical = 4.dp)
            .border(2.dp, border, RoundedCornerShape(14.dp))
            .clickable(onClick = onClick),
        shape = RoundedCornerShape(14.dp),
        colors = CardDefaults.cardColors(containerColor = bg)
    ) {
        Row(Modifier.padding(12.dp), verticalAlignment = Alignment.CenterVertically) {
            Column(Modifier.weight(1f)) {
                Text(et, fontWeight = FontWeight.Bold, fontSize = 15.sp, color = scheme.onSurface)
                Text(ru, fontSize = 12.sp, color = scheme.onSurface.copy(alpha = 0.65f))
            }
            SpeakButton(onClick = onSpeak)
        }
    }
}

@Composable
private fun ResultView(vm: GameViewModel, onExit: () -> Unit) {
    val r = vm.lastResult ?: return
    Column(
        Modifier.fillMaxSize().background(MaterialTheme.colorScheme.background)
            .verticalScroll(rememberScrollState()).padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Spacer(Modifier.height(24.dp))
        Text(if (r.perfect) "🏆" else "🏁", fontSize = 64.sp)
        Text(if (r.perfect) "Доставка без ошибок!" else "Доставка выполнена",
            fontWeight = FontWeight.Bold, fontSize = 24.sp,
            color = MaterialTheme.colorScheme.onBackground)
        Text("${r.order.restaurant} → ${r.order.customer}", fontSize = 14.sp,
            color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.7f))
        Spacer(Modifier.height(20.dp))
        Card(Modifier.fillMaxWidth(), shape = RoundedCornerShape(20.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface)) {
            Column(Modifier.padding(20.dp), verticalArrangement = Arrangement.spacedBy(10.dp)) {
                ResultRow("Верных ответов", "${r.steps}")
                ResultRow("Ошибок", if (r.mistakes == 0) "нет 🎯" else "${r.mistakes}")
                ResultRow("Оплата доставки", "%.2f €".format(r.order.payout))
                ResultRow("Чаевые за качество", if (r.tip > 0) "+${r.tipStr}" else "—")
                ResultRow("Опыт", "+${r.xp} XP")
                Box(Modifier.fillMaxWidth().height(1.dp).background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.1f)))
                ResultRow("Итого заработано", r.earnedStr, highlight = true)
            }
        }
        Spacer(Modifier.height(24.dp))
        Button(onClick = { vm.closeResult(); onExit() },
            modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(14.dp)) {
            Text("Дальше работать 🛵", fontWeight = FontWeight.Bold)
        }
    }
}

@Composable
private fun ResultRow(label: String, value: String, highlight: Boolean = false) {
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
        Text(label, fontSize = if (highlight) 16.sp else 14.sp,
            fontWeight = if (highlight) FontWeight.Bold else FontWeight.Normal,
            color = MaterialTheme.colorScheme.onSurface)
        Text(value, fontSize = if (highlight) 18.sp else 14.sp, fontWeight = FontWeight.Bold,
            color = if (highlight) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurface)
    }
}

File: kuller_flutter/lib/screens/order_screen.dart
import 'package:flutter/material.dart';

import '../game_vm.dart';
import '../models.dart';
import '../widgets.dart';

class OrderScreen extends StatefulWidget {
  final GameViewModel vm;
  const OrderScreen({super.key, required this.vm});

  @override
  State<OrderScreen> createState() => _OrderScreenState();
}

class _OrderScreenState extends State<OrderScreen> {
  final _listCtrl = ScrollController();
  int _lastAutoScrollKey = -1;

  GameViewModel get vm => widget.vm;

  void _autoScroll(int key) {
    if (key == _lastAutoScrollKey) return;
    _lastAutoScrollKey = key;
    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (!_listCtrl.hasClients) return;
      _listCtrl.animateTo(
        _listCtrl.position.maxScrollExtent,
        duration: const Duration(milliseconds: 250),
        curve: Curves.easeOut,
      );
    });
  }

  @override
  void dispose() {
    _listCtrl.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ListenableBuilder(
      listenable: vm,
      builder: (context, _) {
        final session = vm.session;
        if (session == null) {
          WidgetsBinding.instance.addPostFrameCallback((_) {
            if (mounted && Navigator.of(context).canPop()) {
              Navigator.of(context).pop();
            }
          });
          return const Scaffold(body: SizedBox.shrink());
        }
        if (session.finished && vm.lastResult != null) {
          return _ResultView(vm: vm);
        }

        final scheme = Theme.of(context).colorScheme;
        final messages = session.messages(session.activeThread);
        final showTyping = session.typing == session.activeThread;
        _autoScroll(Object.hash(
          messages.length,
          showTyping,
          session.awaiting,
          session.atEnd,
          session.activeThread,
          session.nav != null,
        ));

        return Scaffold(
          backgroundColor: scheme.surfaceContainerLowest,
          body: SafeArea(
            child: Column(
              children: [
                Padding(
                  padding:
                      const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                  child: Row(
                    children: [
                      TextButton.icon(
                        onPressed: () {
                          vm.cancelOrder();
                        },
                        icon: const Icon(Icons.close, size: 18),
                        label: const Text('Отмена'),
                      ),
                      Expanded(
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(
                                '${session.order.restaurant} → ${session.order.customer}',
                                style: TextStyle(
                                    fontWeight: FontWeight.bold,
                                    fontSize: 14,
                                    color: scheme.onSurface)),
                            Text(session.order.address,
                                style: TextStyle(
                                    fontSize: 11,
                                    color: scheme.onSurface.withOpacity(0.6))),
                          ],
                        ),
                      ),
                      IconButton(
                        tooltip: 'Скрыть/показать русский перевод',
                        onPressed: vm.toggleHideRu,
                        icon: Icon(
                            vm.hideRu
                                ? Icons.visibility_off
                                : Icons.visibility,
                            size: 20),
                      ),
                    ],
                  ),
                ),
                LinearProgressIndicator(value: session.progress, minHeight: 5),

                // Вкладки-чаты
                if (session.tabs.length > 1)
                  SingleChildScrollView(
                    scrollDirection: Axis.horizontal,
                    padding:
                        const EdgeInsets.symmetric(horizontal: 8, vertical: 6),
                    child: Row(
                      children: [
                        for (final t in session.tabs)
                          Padding(
                            padding: const EdgeInsets.only(right: 8),
                            child: FilterChip(
                              selected: t == session.activeThread,
                              onSelected: (_) => vm.setActiveThread(t),
                              label: Text(threadLabel(t) +
                                  (session.unread.contains(t) ? '  🔴' : '')),
                            ),
                          ),
                      ],
                    ),
                  ),

                // Лента активного чата
                Expanded(
                  child: ListView(
                    controller: _listCtrl,
                    padding: const EdgeInsets.symmetric(
                        horizontal: 12, vertical: 8),
                    children: [
                      for (final msg in messages)
                        Padding(
                          padding: const EdgeInsets.only(bottom: 8),
                          child: _ChatBubble(
                            msg: msg,
                            displayRu: vm.hideRu ? '' : vm.maskRu(msg.ru),
                            onSpeak: () =>
                                vm.speak(msg.et, msg.fromCourier, msg.thread),
                          ),
                        ),
                      if (showTyping)
                        _TypingBubble(
                            label: threadLabel(session.activeThread)),
                    ],
                  ),
                ),

                // Нижняя панель
                Material(
                  color: scheme.surface,
                  elevation: 8,
                  child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: _bottomPanel(context, session),
                  ),
                ),
              ],
            ),
          ),
        );
      },
    );
  }

  Widget _bottomPanel(BuildContext context, OrderSession session) {
    final scheme = Theme.of(context).colorScheme;
    final ask = session.currentAsk;
    final nav = session.nav;

    if (session.atEnd) {
      return SizedBox(
        width: double.infinity,
        child: FilledButton(
          onPressed: vm.finishDelivery,
          style: FilledButton.styleFrom(
              shape:
                  RoundedRectangleBorder(borderRadius: BorderRadius.circular(14))),
          child: const Text('Завершить доставку 🏁',
              style: TextStyle(fontWeight: FontWeight.bold)),
        ),
      );
    }
    if (nav != null) {
      return Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: Text(nav.promptRu,
                style: TextStyle(
                    fontWeight: FontWeight.bold,
                    fontSize: 15,
                    color: scheme.onSurface)),
          ),
          for (final opt in nav.options)
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 4),
              child: SizedBox(
                width: double.infinity,
                child: FilledButton(
                  onPressed: () => vm.navChoose(opt),
                  style: FilledButton.styleFrom(
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(14))),
                  child: Text(opt.labelRu,
                      style: const TextStyle(fontWeight: FontWeight.bold)),
                ),
              ),
            ),
        ],
      );
    }
    if (session.awaiting &&
        ask != null &&
        ask.thread == session.activeThread) {
      // Прокручиваемые варианты + ЗАКРЕПЛЁННАЯ кнопка «Отправить»
      return Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          ConstrainedBox(
            constraints: const BoxConstraints(maxHeight: 270),
            child: SingleChildScrollView(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(ask.promptRu,
                      style: TextStyle(
                          fontWeight: FontWeight.bold,
                          fontSize: 14,
                          color: scheme.onSurface)),
                  Padding(
                    padding: const EdgeInsets.only(top: 2, bottom: 6),
                    child: Text(
                        ask.courier
                            ? '🛵 Выберите, что сказать:'
                            : '🛵 Выберите ответ:',
                        style: TextStyle(
                            fontSize: 12,
                            fontWeight: FontWeight.bold,
                            color: scheme.secondary)),
                  ),
                  for (var i = 0; i < ask.choices.length; i++)
                    _AnswerOption(
                      et: ask.choices[i].et,
                      ru: vm.hideRu ? '' : ask.choices[i].ru,
                      selected: session.selected == i,
                      onTap: () => vm.select(i),
                      onSpeak: () =>
                          vm.speak(ask.choices[i].et, true, ask.thread),
                    ),
                ],
              ),
            ),
          ),
          const SizedBox(height: 8),
          FilledButton.icon(
            onPressed: session.selected != null ? vm.confirm : null,
            icon: const Icon(Icons.send, size: 18),
            label: const Text('Отправить',
                style: TextStyle(fontWeight: FontWeight.bold)),
            style: FilledButton.styleFrom(
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(14))),
          ),
        ],
      );
    }
    if (session.awaiting && ask != null) {
      return Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: Text('Ответьте в чате «${threadLabel(ask.thread)}»',
                style: TextStyle(fontSize: 13, color: scheme.onSurface)),
          ),
          OutlinedButton(
            onPressed: () => vm.setActiveThread(ask.thread),
            style: OutlinedButton.styleFrom(
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(14))),
            child: Text('Открыть «${threadLabel(ask.thread)}»'),
          ),
        ],
      );
    }
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 12),
      child: Text('💬 …',
          style:
              TextStyle(fontSize: 16, color: scheme.onSurface.withOpacity(0.5))),
    );
  }
}

class _ChatBubble extends StatelessWidget {
  final ChatMsg msg;
  final String displayRu;
  final VoidCallback onSpeak;
  const _ChatBubble(
      {required this.msg, required this.displayRu, required this.onSpeak});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final courier = msg.fromCourier;
    final bubbleColor = courier ? scheme.primary : scheme.surfaceContainerHighest;
    final onBubble = courier ? scheme.onPrimary : scheme.onSurface;
    final radius = BorderRadius.only(
      topLeft: const Radius.circular(18),
      topRight: const Radius.circular(18),
      bottomLeft: Radius.circular(courier ? 18 : 4),
      bottomRight: Radius.circular(courier ? 4 : 18),
    );
    return Row(
      mainAxisAlignment:
          courier ? MainAxisAlignment.end : MainAxisAlignment.start,
      children: [
        ConstrainedBox(
          constraints: const BoxConstraints(maxWidth: 300),
          child: Column(
            crossAxisAlignment:
                courier ? CrossAxisAlignment.end : CrossAxisAlignment.start,
            children: [
              Padding(
                padding: const EdgeInsets.only(left: 6, right: 6, bottom: 2),
                child: Text(msg.label,
                    style: TextStyle(
                        fontSize: 11,
                        fontWeight: FontWeight.bold,
                        color: scheme.onSurface.withOpacity(0.6))),
              ),
              Material(
                color: bubbleColor,
                borderRadius: radius,
                child: Padding(
                  padding: const EdgeInsets.only(
                      left: 12, right: 4, top: 8, bottom: 8),
                  child: Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Flexible(
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(msg.et,
                                style: TextStyle(
                                    fontWeight: FontWeight.bold,
                                    fontSize: 16,
                                    color: onBubble)),
                            if (displayRu.trim().isNotEmpty)
                              Text(displayRu,
                                  style: TextStyle(
                                      fontSize: 13,
                                      color: onBubble.withOpacity(0.8))),
                          ],
                        ),
                      ),
                      const SizedBox(width: 4),
                      SpeakButton(
                          onPressed: onSpeak,
                          tint: courier ? scheme.onPrimary : scheme.primary),
                    ],
                  ),
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }
}

class _TypingBubble extends StatelessWidget {
  final String label;
  const _TypingBubble({required this.label});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 6, right: 6, bottom: 2),
          child: Text(label,
              style: TextStyle(
                  fontSize: 11,
                  fontWeight: FontWeight.bold,
                  color: scheme.onSurface.withOpacity(0.6))),
        ),
        Material(
          color: scheme.surfaceContainerHighest,
          borderRadius: const BorderRadius.only(
            topLeft: Radius.circular(18),
            topRight: Radius.circular(18),
            bottomLeft: Radius.circular(4),
            bottomRight: Radius.circular(18),
          ),
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 10),
            child: Text('печатает •••',
                style: TextStyle(
                    fontSize: 14, color: scheme.onSurface.withOpacity(0.6))),
          ),
        ),
      ],
    );
  }
}

class _AnswerOption extends StatelessWidget {
  final String et;
  final String ru;
  final bool selected;
  final VoidCallback onTap;
  final VoidCallback onSpeak;
  const _AnswerOption({
    required this.et,
    required this.ru,
    required this.selected,
    required this.onTap,
    required this.onSpeak,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final border =
        selected ? scheme.primary : scheme.onSurface.withOpacity(0.12);
    final bg = selected ? scheme.primary.withOpacity(0.08) : scheme.surface;
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 4),
      child: Material(
        color: bg,
        borderRadius: BorderRadius.circular(14),
        child: InkWell(
          borderRadius: BorderRadius.circular(14),
          onTap: onTap,
          child: Container(
            decoration: BoxDecoration(
              border: Border.all(color: border, width: 2),
              borderRadius: BorderRadius.circular(14),
            ),
            padding: const EdgeInsets.all(12),
            child: Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(et,
                          style: TextStyle(
                              fontWeight: FontWeight.bold,
                              fontSize: 15,
                              color: scheme.onSurface)),
                      if (ru.trim().isNotEmpty)
                        Text(ru,
                            style: TextStyle(
                                fontSize: 12,
                                color: scheme.onSurface.withOpacity(0.65))),
                    ],
                  ),
                ),
                SpeakButton(onPressed: onSpeak),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class _ResultView extends StatelessWidget {
  final GameViewModel vm;
  const _ResultView({required this.vm});

  @override
  Widget build(BuildContext context) {
    final r = vm.lastResult;
    if (r == null) return const Scaffold(body: SizedBox.shrink());
    final scheme = Theme.of(context).colorScheme;
    return Scaffold(
      backgroundColor: scheme.surfaceContainerLowest,
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(24),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              const SizedBox(height: 24),
              Text(r.perfect ? '🏆' : '🏁',
                  textAlign: TextAlign.center,
                  style: const TextStyle(fontSize: 64)),
              Text(r.perfect ? 'Доставка без ошибок!' : 'Доставка выполнена',
                  textAlign: TextAlign.center,
                  style: TextStyle(
                      fontWeight: FontWeight.bold,
                      fontSize: 24,
                      color: scheme.onSurface)),
              Text('${r.order.restaurant} → ${r.order.customer}',
                  textAlign: TextAlign.center,
                  style: TextStyle(
                      fontSize: 14, color: scheme.onSurface.withOpacity(0.7))),
              const SizedBox(height: 20),
              Card(
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(20)),
                color: scheme.surface,
                child: Padding(
                  padding: const EdgeInsets.all(20),
                  child: Column(
                    children: [
                      _resultRow(context, 'Верных ответов', '${r.steps}'),
                      _resultRow(context, 'Ошибок',
                          r.mistakes == 0 ? 'нет 🎯' : '${r.mistakes}'),
                      _resultRow(context, 'Оплата доставки',
                          '${r.order.payout.toStringAsFixed(2)} €'),
                      _resultRow(context, 'Чаевые за качество',
                          r.tip > 0 ? '+${r.tipStr}' : '—'),
                      _resultRow(context, 'Опыт', '+${r.xp} XP'),
                      Divider(color: scheme.onSurface.withOpacity(0.1)),
                      _resultRow(context, 'Итого заработано', r.earnedStr,
                          highlight: true),
                    ],
                  ),
                ),
              ),
              const SizedBox(height: 24),
              FilledButton(
                onPressed: () {
                  vm.closeResult();
                  Navigator.of(context).pop();
                },
                style: FilledButton.styleFrom(
                    shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(14))),
                child: const Text('Дальше работать 🛵',
                    style: TextStyle(fontWeight: FontWeight.bold)),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _resultRow(BuildContext context, String label, String value,
      {bool highlight = false}) {
    final scheme = Theme.of(context).colorScheme;
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 5),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Text(label,
              style: TextStyle(
                  fontSize: highlight ? 16 : 14,
                  fontWeight: highlight ? FontWeight.bold : FontWeight.normal,
                  color: scheme.onSurface)),
          Text(value,
              style: TextStyle(
                  fontSize: highlight ? 18 : 14,
                  fontWeight: FontWeight.bold,
                  color: highlight ? scheme.primary : scheme.onSurface)),
        ],
      ),
    );
  }
}
